Generate changelog for 3.0.0-beta1 (#9381)
tarantool/tarantool: changelogs/3.0.0-beta1.md

## Functionality added or changed

### Core

- **[Breaking change]** Disabled DDL operations with an old system schema.
  Now you have to run `box.schema.upgrade()` before you can execute any DDL
  operations (gh-7149).
* Now `index_object.parts` contains the following methods, similar to the
  `key_def` Lua module: `extract_key()`, `compare()`, `compare_with_key()`,
  `merge()` (gh-7356).
* Introduced the `box.watch_once()` function to get the value currently
  associated with a notification key on the local instance without subscribing
  to future changes (gh-6493).
* Introduced the new `IPROTO_WATCH_ONCE` request to get the value currently
  associated with a notification key on a remote instance without subscribing
  to future changes (gh-6493).
* Introduced the `conn:watch_once()` net.box connection method to get the value
  currently associated with a notification key on a remote instance without
  subscribing to future changes. The new method is implemented using the
  `IPROTO_WATCH_ONCE` request type (gh-6493).
* Introduced the default field values in the space format (gh-8157).
* Added support for functional variant of the default field values (gh-8609).

> * box.schema can now be accessed before box.cfg().

Reviewer: ```suggestion
* `box.schema` can now be accessed before `box.cfg()`.
```

* Added the ability to set the `bootstrap_leader` configuration option to the
  instance name of the desired bootstrap leader:
  ```lua
    box.cfg{
        bootstrap_strategy = 'config',
        bootstrap_leader = 'leader-name',
        replication = {
            ...
        },
        ...
    }
  ```
  (gh-7999, gh-8539).

### LuaJIT

* LuaJIT now can handle exceptions on traces (gh-7745).

### Lua

* **[Breaking change]** Added the new `varbinary` type to Lua. An object of
  this type is similar to a plain string but encoded in MsgPack as `MP_BIN` so
  it can be used for storing binary blobs in the database. This also works the
  other way round: data fields stored as `MP_BIN` are now decoded in Lua as
  varbinary objects, not as plain strings, as they used to be. Since the latter
  may cause compatibility issues, the new compat option `binary_data_decoding`
  was introduced to revert the built-in decoder to the old behavior (gh-1629).
* Introduced the new Lua module `trigger`. This module enables managing and
  calling triggers stored in trigger registry (gh-8656).
* Triggers from `space_object`, `box.session`, and `box.ctl` were moved to
  the trigger registry (gh-8657).

> * Introduced a new event 'tarantool.trigger.on_change' in the trigger registry.

Reviewer: ```suggestion
* Introduced a new event `'tarantool.trigger.on_change'` in the trigger registry.
```

  (gh-7986).

#### Netbox

* Introduced the new `net.box` module function `from_fd` for creating a new
  connection from a socket file descriptor number (gh-8984).

#### Popen

* Introduced new option `inherit_fds` for `popen.new`. The option takes
  an array of file descriptor numbers that should be left open in the child
  process (gh-8926).

#### Socket

* Introduced new socket functions `socket.socketpair`, `socket.from_fd`, and
  `socket:detach` (gh-8927).

#### Xlog

* Introduced the `xlog.meta()` method for reading a meta block from an xlog
  file.

### SQL

* Introduced the `SHOW CREATE TABLE` statement (gh-8098).
* **[Breaking change]** The `DROP CONSTRAINT` statement has been
  improved (gh-9112).
* **[Breaking change]** Names in SQL are now case sensitive. To support backward
  compatibility, a second lookup using a name normalized using the old rules is
  added (gh-4467).
* `SHOW CREATE TABLE` no longer supports the DEFAULT clause (gh-8793).
* `SQL_EXPR` functions can now be set as a default value (gh-8793).
* A literal set as the default value can no longer have a `+` or `-` sign unless
  the literal is numeric (gh-8793).

> * **[Breaking change]** SQL now uses the BOX mechanism for default

Reviewer: ```suggestion
* **[Breaking change]** SQL now uses the `box` mechanism for default
```

  values (gh-8793).

### App

* Added the new `--force-recovery` CLI option (gh-8876).

### Box

* Non-system spaces in a snapshot file are now sorted by their IDs.
  As before, all non-system spaces are stored after system spaces (gh-7954).
* **[Breaking change]** Dropped support for `IPROTO_CALL16` from `net.box`
  (i.e., the `call_16` option and the `call_16` method for `net.box`
  connections) (gh-8636).
* Added support for square brackets in procedure resolution for Lua calls
  (gh-8604).
* Added support for accepting IPROTO requests with a space or index name instead
  of an identifier (gh-8146).
* Added a new `box.info` parameter `hostname` (gh-8605).
* Added the `box.cfg.memtx_sort_threads` parameter that specifies the number of
  threads used to sort indexes keys on loading a memtx database. OpenMP is
  not used to sort keys anymore (gh-3389).
* Changed the output of the `fselect` index method from a table of strings to

>   a multi-line string and dropped the `print` and `use_nbsp` options. With the

Reviewer: ```suggestion
  a multiline string and dropped the `print` and `use_nbsp` options. With the
```

>   `yaml_pretty_multiline` compat option enabled by default, multi-line strings

Reviewer: ```suggestion
  `yaml_pretty_multiline` compat option enabled by default, multiline strings
```

  `box.schema.user.grant`. Granting the `'execute'` privilege on `lua_eval`
  allows the user to execute an arbitrary Lua expression with the
  `IPROTO_EVAL` request. Granting the `'execute'` privilege on `lua_call`
  allows the user to execute any global user-defined Lua function with
  the `IPROTO_CALL` request (gh-8803).
* **[Breaking change]** Introduced the new `sql` object type for
  `box.schema.user.grant`. Now only users with the `'execute'` privilege
  granted on `sql` or `universe` can execute SQL expressions with the
  `IPROTO_EXECUTE` or `IPROTO_PREPARE` requests. To revert to the old behavior
  (no SQL access checks), use the `sql_priv` compat option (gh-8803).
* The maximum length of `box.cfg{}` string parameters is now 512 instead of 256.
* Added a new flag `tarantool.build.asan` that shows whether build
  flag `ENABLE_ASAN` is set.
* Added the `leader_name` field to `box.info.election` (gh-8931).
* Now it is possible to specify the sort order for each part of an index
  (gh-5529).
* Added support for sending tuple formats in IPROTO responses. Added a
  `box_tuple_extension` backward compatibility option to disable sending

>   tuple formats in responses to IPROTO call and eval requests (gh-8146).

Reviewer: ```suggestion
  tuple formats in responses to IPROTO_CALL and IPROTO_EVAL requests (gh-8146).
```

> * Added support for sending tuple formats in IPROTO call and eval request

Reviewer: ```suggestion
* Added support for sending tuple formats in IPROTO_CALL and IPROTO_EVAL request
```

  arguments. `box_tuple_extension` backward compatibility option can be used to

>   disable receiving tuple formats in IPROTO call and eval request arguments

Reviewer: ```suggestion
  disable receiving tuple formats in IPROTO_CALL and IPROTO_EVAL request arguments
```

  supported by credentials (gh-8967).
* Now a password hash (and salt) will be regenerated for users managed
  in the configuration file if `security.auth_type` differs from a user's
  `auth_type` (gh-8967).
* Introduced validation for replicaset_name/uuid and instance_name/uuid
  mismatches before the recovery process when Tarantool is configured via
  a YAML file or etcd.
* If Tarantool is configured via a YAML file or etcd, then instance and replica
  set names are automatically set when possible (gh-8978).
* Added support for the `{{ replicaset_name }}` and `{{ group_name }}`
  templates in addition to the existing `{{ instance_name }}` (gh-8862).

### Console

* Now multiline commands can be used by setting continuation marker
  using the `\set continuation` command.

### Debugger

* Introduced readline support to tarantool debugger (gh-7738).

### Fiber

* Exported `fiber_set_name_n`, `fiber_name`, `fiber_id`, `fiber_csw` and
  `fiber_find` into the public C API and usable via FFI as well.
* Make `fiber_set_joinable`, `fiber_set_ctx` and `fiber_get_ctx`
  treat the NULL argument as the current fiber.

### Space

> * Introduces space type: a new space definition field.
> * Introduces the fully temporary space type. It is the same as data-temporary
>   but also has temporary metadata. Temporary spaces can now be created in
>   read_only mode, they disappear after server restart and don't exist on
>   replicas (gh-8323).

Reviewer: ```suggestion
* Introduced space type: a new space definition field.
* Introduced the fully temporary space type. It is the same as data-temporary
  but also has temporary metadata. Temporary spaces can now be created in
  read-only mode, they disappear after server restart and don't exist on
  replicas (gh-8323).
```
